Created Tetris clone "Reblox"

Stephen-Seo opened the pull request:
The main Makefile in the Redox OS repository will need to be updated to build this game.

Stephen-Seo committed:
Made tetris clone "Reblox"

Game logic is mostly done in grid.rs.
Randomness is provided by pseudo_random.rs, which isn't perfect.
The rest is in main.rs

Stephen-Seo committed:
Update README.md

EGhiorzi commented:
Are you sure the Makefile has to be changed? I thought other games didn't need that…

Stephen-Seo commented:
@EGhiorzi Take a look at line 236 in the main Makefile for redox-os. The other games are specified there. If you just build and run redox-os without any changes, the game reblox is not available.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    (START_BLOCK_SPEED_NANOSEC - END_BLOCK_SPEED_NANOSEC) / 255;
+const LINES_TO_CLEAR_TO_LVL_UP: u8 = 10;
+
+pub fn p2xy(pos: u8) -> (u8, u8) {
```
It is generally considered an anti-pattern to use lonely functions for convertions. You should instead define them on the type itself.

For example, create a type for `pos` and impl `Into<(u8, u8)>`.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    (pos % GRID_WIDTH, pos / GRID_WIDTH)
+}
+
+pub fn xy2p(x: u8, y: u8) -> u8 {
```
Same here.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    x + y * GRID_WIDTH
+}
+
+pub fn pr2piece(pos: u8, rotation: u8, block_type: BlockType, grid_width: u8) -> [u8;4] {
```
Here too.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+}
+
+pub fn pr2piece(pos: u8, rotation: u8, block_type: BlockType, grid_width: u8) -> [u8;4] {
+    match rotation {
```
Perhaps generalize this form instead of one for each rotation (just multiply by `i^rotation`).

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    falling_rot: u8,
+    next_falling_type: BlockType,
+    next_falling_rot: u8,
+    pub rng: PseudoRandom,
```
Why not `Randomizer` from libextra?

ticki replied:
Why not `Randomizer` from libextra?

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+
+impl Grid {
+    pub fn new() -> Grid {
+        Grid {
```
This struct is like 90% `Default::default()`. Derive `Default` and simply override the respecitve field through the `.. Grid::default()` syntax.

ticki replied:
This struct is like 90% `Default::default()`. Derive `Default` and simply override the respecitve field through the `.. Grid::default()` syntax.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+            self.grid[i as usize] = BlockType::NONE;
+        }
+        self.falling_type = BlockType::NONE;
+        match self.rng.next() % 7 {
```
This really should be some conversion method.

ticki replied:
This really should be some conversion method.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+        }
+    }
+
+    pub fn drop(&mut self) {
```
Rename this to `fall`? `drop` might be confusing given it isn't a destructor.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    S,
+    T,
+    Z,
+    GARBAGE,
```
This isn't following naming convention. Variants are type-like named.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+        self.falling_pos = xy2p(GRID_WIDTH / 2, GRID_HEIGHT - 1);
+        self.falling_type = self.next_falling_type;
+        self.falling_rot = self.next_falling_rot;
+        match self.rng.next() % 7 {
```
DRY!

ticki commented on src/reblox/pseudo_random.rs:
```diff
@@ -0,0 +1,16 @@
+        PseudoRandom { seed: 0 }
+    }
+
+    pub fn next(&mut self) -> u64 {
```
Perhaps use the PRNG provided by libextra (see `minesweeper`).

On a more technical note: RNGs generally applies some bijective function to the result to avoid patterns (e.g. XOR with some number, wrapping multiplication with a prime, wrapping addition, and so on).

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    }
+}
+
+/// (0, 0) is bottom left
```
You should add proper docs to this struct and its field: some of it might not be obvious at first.

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+    pub grid_size: u8,
+    falling_type: BlockType,
+    falling_pos: u8,
+    falling_rot: u8,
```
Coudln't you actually just store the brick itself as an array of cells?

ticki commented on src/reblox/grid.rs:
```diff
@@ -0,0 +1,595 @@
+        }
+    }
+
+    pub fn rotate_clockwise(&mut self) {
```
This function seems like a hack?

ticki commented:
Looks very good! I have left quite a few comments, mostly about style and conventions.

Stephen-Seo committed:
Fixed Reblox according to ticki's suggestions

Fixed DRY issue.

Fixed some lone functions issue.

Replaced pr2piece with struct.
Avoids anti-pattern lonely functions.

Replaced pseudo_random with libextra's Randomizer.

Fixed naming convention of enum variants.

Renamed member function from "drop" to "fall" to avoid confusion.

Added documentation to Grid struct.

Stephen-Seo commented:
I made changes according to your recommendations with some exceptions:
* Did not #[derive(Default)] for struct Grid because it has an array and I wasn't able to get it working with that array.
* Falling blocks are not stored as array of cells but rather stored in Grid::grid. This is done purposefully so that outside this module, one would only need to look at the Grid::grid to draw the board. The only thing not drawn from Grid::grid is the next falling block.

ticki commented:
Great!
